### ISSUE TITLE
Use decorators to define steps in Page Object Models

### ISSUE THREAD

**Issue by vitalets** [vitalets/playwright-bdd#35]
Imagine I use `TodoPage` fixture (page object model) from [Playwright docs](https://playwright.dev/docs/test-fixtures#creating-a-fixture):
```ts
// fixtures.ts

export class TodoPage {
  private readonly inputBox: Locator;
  private readonly todoItems: Locator;

  constructor(public readonly page: Page) {
    this.inputBox = this.page.locator('input.new-todo');
    this.todoItems = this.page.getByTestId('todo-item');
  }

  async goto() {
    await this.page.goto('https://demo.playwright.dev/todomvc/');
  }

  async addToDo(text: string) {
    await this.inputBox.fill(text);
    await this.inputBox.press('Enter');
  }
}
```

Typical BDD step definitions for `todoPage` will be:
```ts
// steps.ts

Given('I am on todo page', async ({ todoPage }) => {
  await todoPage.goto();
});

When('I add todo {string}', async ({ todoPage }, text: string) => {
  await todoPage.addToDo(text);
});
```

For me such step definitions look like a boilerplate code. It would be more convenient to just mark `todoPage` methods with step text - the same way as it's done in [Cucumber Java](https://cucumber.io/docs/cucumber/step-definitions/?lang=java). Having new JS [decorators](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) syntax it should be possible:
```ts
// fixtures.ts

export class TodoPage {
  private readonly inputBox: Locator;
  private readonly todoItems: Locator;

  constructor(public readonly page: Page) {
    this.inputBox = this.page.locator('input.new-todo');
    this.todoItems = this.page.getByTestId('todo-item');
  }

  @Given('I am on todo page')
  async goto() {
    await this.page.goto('https://demo.playwright.dev/todomvc/');
  }

  @When('I add todo {string}')
  async addToDo(text: string) {
    await this.inputBox.fill(text);
    await this.inputBox.press('Enter');
  }
}
```
In that case there is no need for additional `steps.ts` file. We just define fixture and write scenarios.

Would appreciate your ideas on that.

**Comment by NikkTod** [comment]
@vitalets This would definitely save us some code typing and linking between feature -> steps -> pages. 
Hope it is not affecting the VSCode BDD extensions with all the bindings :)

**Comment by jzaratei** [comment]
That could be interesting, but what if I want to make use of multiples pages objects in a single step, for example:
```
When("I delete an old project", async ({ ProjectsPage, deleteProjectPage }) => {
  oldProjectName = await ProjectsPage.getLastProjectName();
  await deleteProjectPage.deleteProject(oldProjectName);
});
```
This implementation could be still be possible?

**Comment by vitalets** [comment]
> That could be interesting, but what if I want to make use of multiples pages objects in a single step, for example:
> 
> ```
> When("I delete an old project", async ({ ProjectsPage, deleteProjectPage }) => {
>   oldProjectName = await ProjectsPage.getLastProjectName();
>   await deleteProjectPage.deleteProject(oldProjectName);
> });
> ```
> 
> This implementation could be still be possible?

Yes, that will be possible as regular `Given / When / Then` functions will be available as well.

Btw, if "delete project" is some link / button on `ProjectsPage` - shouldn't it be part of `ProjectsPage`?

**Comment by vitalets** [comment]
Released in v4.0.0. [Docs](https://github.com/vitalets/playwright-bdd#decorators).

I would appreciate if try it and share your feedback :) 
Personally I've moved one of my projects to BDD decorators and I found this syntax very concise.

**Comment by konkakaka** [comment]
@vitalets which extension on Visual Studio Code on Mac that would suggest the steps while typing existing defined step on feature file? I tried `Cucumber (Gherkin) Full Support`, and others, but it doesn't seem work, thank you!

**Comment by vitalets** [comment]
@konkakaka `Cucumber (Gherkin) Full Support` works for me. Could you share more details, especially .vscode/settings.json?
